From 8112d18b515d737475e502470f088b37b4d45dee Mon Sep 17 00:00:00 2001
From: Johan Mossberg <johan.mossberg@intel.com>
Date: Thu, 11 Dec 2014 10:46:30 +0100
Subject: [PATCH 232/236] OV5693: Emulate R_PWN (PWDNB) tied to DOVDD in SW

According to the OV5693 DS PWDND should be tied to DOVDD which is not
the case on ECS. By emulating this behavior in SW we get rid of the
"lost arbitration" errors that sometimes occur on ECS. Arbitration
problems should not occur on ECS I2C1 as there is only one master.
The theory is that OV5693 or AD5823 is misbehaving during startup,
temporarily jamming the I2C bus.

Change-Id: Idbb69a00cd054e870b92c8561238fdd0815cfe27
Tracked-On: GMINL-1773
Signed-off-by: Johan Mossberg <johan.mossberg@intel.com>
Reviewed-on: https://icggerrit.ir.intel.com/32385
Reviewed-by: Jing Tao <jing.tao@intel.com>
Approver: Wei Tao <wei.tao@intel.com>
Reviewed-by: Wei Tao <wei.tao@intel.com>
---
 .../camera/drivers/media/i2c/ov5693-ecs/ov5693.c      | 19 ++++++++++---------
 1 file changed, 10 insertions(+), 9 deletions(-)

diff --git a/drivers/external_drivers/camera/drivers/media/i2c/ov5693-ecs/ov5693.c b/drivers/external_drivers/camera/drivers/media/i2c/ov5693-ecs/ov5693.c
index 9f5b17a..45c7917 100644
--- a/drivers/external_drivers/camera/drivers/media/i2c/ov5693-ecs/ov5693.c
+++ b/drivers/external_drivers/camera/drivers/media/i2c/ov5693-ecs/ov5693.c
@@ -1328,14 +1328,23 @@ static int power_ctrl(struct v4l2_subdev *sd, bool flag)
 	if (dev->platform_data->power_ctrl)
 		return dev->platform_data->power_ctrl(sd, flag);
 
+	/* This driver assumes "internal DVDD, PWDNB tied to DOVDD".
+	 * In this set up only gpio0 (XSHUTDN) should be available
+	 * but in some products (for example ECS) gpio1 (PWDNB) is
+	 * also available. If gpio1 is available we emulate it being
+	 * tied to DOVDD here. */
 	if (flag) {
 		ret = dev->platform_data->v2p8_ctrl(sd, 1);
+		dev->platform_data->gpio1_ctrl(sd, 1);
 		if (ret == 0) {
 			ret = dev->platform_data->v1p8_ctrl(sd, 1);
-			if (ret)
+			if (ret) {
+				dev->platform_data->gpio1_ctrl(sd, 0);
 				ret = dev->platform_data->v2p8_ctrl(sd, 0);
+			}
 		}
 	} else {
+		dev->platform_data->gpio1_ctrl(sd, 0);
 		ret = dev->platform_data->v1p8_ctrl(sd, 0);
 		ret |= dev->platform_data->v2p8_ctrl(sd, 0);
 	}
@@ -1357,14 +1366,6 @@ static int gpio_ctrl(struct v4l2_subdev *sd, bool flag)
 
 	ret = dev->platform_data->gpio0_ctrl(sd, flag);
 
-	/* The OV5693 has two enable inputs: XSHUTDN and RESETB, both
-	 * are active low, both must be high to enable the device.
-	 * And they can be enabled in either order.  The datasheet
-	 * even suggests that one be tied high, and some modules do
-	 * that.  Basically: allow the second GPIO to be missing in
-	 * the DSDT and ignore an error here. */
-	dev->platform_data->gpio1_ctrl(sd, flag);
-
 	return ret;
 }
 
-- 
1.9.1

